Add 2025-02-16-newpipe-0.27.6-rewrite-team-states blogpost (#388)
TeamNewPipe/website: _posts/2025-02-16-newpipe-0.27.6-rewrite-team-states.md

</figure>

## The refactor

As explained above, in mid 2024 we finally started working on the awaited refactor. For reference, the original announcement is [here](https://github.com/TeamNewPipe/NewPipe/discussions/10118), the **project board** with the issues to tackle is [here](https://github.com/orgs/TeamNewPipe/projects/18/views/1). We chose a mix between just *refactoring* (when the current code is already good enough) and fully *rewriting* (for UI components and for the player). The refactor is currently happening on the `refactor` branch of the NewPipe repository, so that the `dev` branch can still be used for hotfix releases and maintenance bugfixes. The following points explain our efforts and objectives, along with **screenshots** and a video.

### Rewriting the player

<figure class="right">
    <video src="{{ site.baseurl }}/img/screenshots/newplayer.webm" controls=""></video>
    <figcaption class="text-center">A showcase of the NewPlayer test app</figcaption>
</figure>

> As explained in [this previous blogpost]({{ site.baseurl }}/blog/pinned/announcement/schabi-contract/), NewPipe's current player is **buggy and mis-structured**. [Schabi was contracted]({{ site.baseurl }}/blog/pinned/announcement/schabi-contract/) to rewrite it from scratch [by the NewPipe e.V.](https://newpipe-ev.de/posts/2024-06-23_3rd_board_meeting.html), and he delivered a great implementation in October 2024. The new player (or rather, *NewPlayer*) uses [media3](https://developer.android.com/media/media3) for playback (a successor of ExoPlayer), and was built in a separate repository to ensure separation of concerns and so that other apps can possibly use it too. You can **read Schabi's announcement** [here](https://github.com/TeamNewPipe/NewPipe/issues/11603), find the source code [here](https://github.com/TeamNewPipe/NewPlayer) and download a test APK from [here](https://github.com/TeamNewPipe/NewPlayer/releases/). Now [the e.V. made a contract](https://newpipe-ev.de/posts/2024-12-03_hiring_contributor.html) with [Profpatsch](https://github.com/Profpatsch), who picked up Schabi's work and is integrating NewPlayer into NewPipe. Let's wish them well in their work!

Author: Fixed on every developer.android.com link

### Breaking change 2: invalid iOS player responses (fixed in 0.27.5)

YouTube started to require a **visitor data** string. It is used to identify in a unique way the session of a user and is returned by YouTube servers the first time a client makes a request. If a visitorData is not provided, YouTube returns a *"Sign in to confirm that you're not a bot"* error.

To fix this issue, [@Theta-Dev](https://github.com/Theta-Dev) opened [a PR](https://github.com/TeamNewPipe/NewPipeExtractor/pull/1262) where a random visitorData is generated every time.

### Breaking change 3: invalid iOS player responses (fixed in 0.27.6)

A couple of weeks ago, YouTube started to **reject randomly generated visitorData**, and so required visitorData generated by their servers. NewPipe's extractor now makes requests to get a visitorData from YouTube for every player request we make, regardless of the client (just in the case this change applies to all clients at some point).

> As YouTube is requiring integrity checks in their iOS client that we are not able to pass for most of formats returned, we removed usage of this client in NewPipe. This unfortunately means that livestreams cannot be rewinded for more than 30 seconds, since the *web* client does not allow to rewind livestreams in HLS manifests for more than this duration. <!-- TODO please check if correct -->

Author: I turned "the iOS client allows to rewind up to 1 hour in most cases, sometimes only 15 minutes" into just "the iOS client allows to rewind up to 1 hour" because the rest is probably too much information, and applied the rest of the suggestion

</figure>

## The refactor

As explained above, in mid 2024 we finally started working on the awaited refactor. For reference, the original announcement is [here](https://github.com/TeamNewPipe/NewPipe/discussions/10118), the **project board** with the issues to tackle is [here](https://github.com/orgs/TeamNewPipe/projects/18/views/1). We chose a mix between just *refactoring* (when the current code is already good enough) and fully *rewriting* (for UI components and for the player). The refactor is currently happening on the `refactor` branch of the NewPipe repository, so that the `dev` branch can still be used for hotfix releases and maintenance bugfixes. The following points explain our efforts and objectives, along with **screenshots** and a video.

### Rewriting the player

<figure class="right">
    <video src="{{ site.baseurl }}/img/screenshots/newplayer.webm" controls=""></video>
    <figcaption class="text-center">A showcase of the NewPlayer test app</figcaption>
</figure>

> As explained in [this previous blogpost]({{ site.baseurl }}/blog/pinned/announcement/schabi-contract/), NewPipe's current player is **buggy and mis-structured**. [Schabi was contracted]({{ site.baseurl }}/blog/pinned/announcement/schabi-contract/) to rewrite it from scratch [by the NewPipe e.V.](https://newpipe-ev.de/posts/2024-06-23_3rd_board_meeting.html), and he delivered a great implementation in October 2024. The new player (or rather, *NewPlayer*) uses [media3](https://developer.android.com/media/media3) for playback (a successor of ExoPlayer), and was built in a separate repository to ensure separation of concerns and so that other apps can possibly use it too. You can **read Schabi's announcement** [here](https://github.com/TeamNewPipe/NewPipe/issues/11603), find the source code [here](https://github.com/TeamNewPipe/NewPlayer) and download a test APK from [here](https://github.com/TeamNewPipe/NewPlayer/releases/). Now [the e.V. made a contract](https://newpipe-ev.de/posts/2024-12-03_hiring_contributor.html) with [Profpatsch](https://github.com/Profpatsch), who picked up Schabi's work and is integrating NewPlayer into NewPipe. Let's wish them well in their work!

Reviewer: ```suggestion
As explained in [this previous blogpost]({{ site.baseurl }}/blog/pinned/announcement/schabi-contract/), NewPipe's current player is **buggy and mis-structured**. [Schabi was contracted]({{ site.baseurl }}/blog/pinned/announcement/schabi-contract/) to rewrite it from scratch [by the NewPipe e.V.](https://newpipe-ev.de/posts/2024-06-23_3rd_board_meeting.html), and he delivered a great implementation in October 2024. The new player (or rather, *NewPlayer*) uses [media3](https://developer.android.com/media/media3) for playback (a successor of ExoPlayer), and was built in a separate repository to ensure separation of concerns and so that other apps can possibly use it too. You can **read Schabi's announcement** [here](https://github.com/TeamNewPipe/NewPipe/issues/11603), find the source code [here](https://github.com/TeamNewPipe/NewPlayer) and download a test APK from [here](https://github.com/TeamNewPipe/NewPlayer/releases/). Now [the e.V. signed a contract](https://newpipe-ev.de/posts/2024-12-03_hiring_contributor.html) with [Profpatsch](https://github.com/Profpatsch), who picked up Schabi's work and is integrating NewPlayer into NewPipe. Let's wish them and us a lot of success with their work!
```

Author: I'm not so sure about "and us", we never mention the team work in this paragraph but just profpatsch, so I don't know

- NewPipe is still a mix of Java and Kotlin, and we would like to complete the **move to Kotlin**. This means making use of Kotlin's language features, like nullability checks, but also switching from RxJava to the built-in coroutines.
- We want to migrate every **UI** component to **[Jetpack Compose](https://developer.android.com/compose)**, a UI toolkit which employs a *declarative* approach as opposed to the default Android XML-based *imperative* approach. It makes development significantly faster and less error-prone, since it allows writing UI components directly in Kotlin and favours having a single source of truth for the app's state.
- We want to embrace the **Model-View-ViewModel** architectural pattern (the standard one for Android apps).
- We want to use Hilt for **dependency injection**, use `DataStore` instead of `SharedPreferences`, modernize the build system, …

### Redesigning the UI (in Material 3)

While migrating the UI to Jetpack Compose, we want to take the opportunity to redesign the most used UI components and make sure they are *easy* to understand, *convenient* to use, *powerful*, and *nice* to see. We also wish to adopt Material 3.

> Users often ask for new features, and when they are told that those features would be in the way of the average user, they ask for **settings toggles** to tune the app's behavior. While having customizability is good, *too much of it is almost equivalent to not having it at all*: if there are pages upon pages of settings, would you really go through and try what each of them does? Moreover, more settings implies more possible configurations, which implies more code to maintain, which implies more bugs.

Reviewer: ```suggestion
Users often ask for new features, and when they are told that those features would be in the way of the average user, they ask for **settings toggles** to tune the app's behavior. While having customizability is good, *too much of it is almost equivalent to not having it at all*: If there are pages upon pages of settings, would you really go through and try what each of them does? Moreover, more settings implies more possible configurations, which implies more code to maintain, which implies more bugs.
```

<span style="clear: right; display: block;"></span>

[@Isira-Seneviratne](https://github.com/Isira-Seneviratne) also provided the initial implementation for the migrated **comment section**, which [@Stypox](https://github.com/Stypox) took up and redesigned, thanks to both! You can easily contribute by fixing a few leftover bugs highlighted in [the comments](https://github.com/TeamNewPipe/NewPipe/pull/11060#pullrequestreview-2427509598) at the bottom of the PR.<br/>*Finally comment replies behave normally!*

<figure class="screenshot-list">
    <img src="{{ site.baseurl }}/img/screenshots/rewrite_2.png"/>
    <img src="{{ site.baseurl }}/img/screenshots/rewrite_3.png"/>
    <figcaption>The revamped comment section with replies shown in a bottom-sheet dialog</figcaption>
</figure>

> [@Isira-Seneviratne](https://github.com/Isira-Seneviratne) and [@Stypox](https://github.com/Stypox) worked (again) together on the reimplementation of the **About screen** (*that's the most useful part of the app, right?*). In particular they made use of the [AboutLibraries](https://github.com/mikepenz/AboutLibraries) Gradle plugin which automatically compiles information on the dependencies which can be accessed at run time, saving the developers of having to hardcode details about the libraries used.

Reviewer: ```suggestion
[@Isira-Seneviratne](https://github.com/Isira-Seneviratne) and [@Stypox](https://github.com/Stypox) worked (again) together on the reimplementation of the **About screen** (*that's the most useful part of the app, right?*). In particular they made use of the [AboutLibraries](https://github.com/mikepenz/AboutLibraries) Gradle plugin which automatically compiles information on the dependencies which can be accessed at run time. This saves the developers from having to hardcode these details about the libraries used in NewPipe.
```

Many PRs opened on the NewPipe repository share a common challenge: while authors often provide a Proof-of-Concept that demonstrates the idea effectively, **refining all details** to finally merge the PR **requires much more effort**, which often falls on the maintainers. The necessary steps to make a PR ready might include *making the UI nice and responsive to screen size, thinking about the user's usecases and seeing if they are fulfilled, ensuring the code is well structured and complies with Android best practices, adding comments and documentation, testing the changes on various devices with various usage patterns*.

The team is also lacking some **triaging manpower** to handle all of the issues and discussions that are opened on the repository. Despite this it is usually easy to identify new issues/regressions or pressing problems thanks to the active community, but it is much harder to distinguish old tickets that are useful from the niche ones.

> I want to end this paragraph with a **call-to-action**: if you have some time and want to help, feel free to get in touch with us [on the Matrix channel](#conclusion-and-contact) and ask us for something to work on, and **we will give you indications**. When opening a Pull Request, make sure your proposal is not just a Proof-of-Concept, but **provide the best code you can write and highlight the areas that need more work**, so the team can review and merge the PR fast. If you want to learn modern Android development (Kotlin, Jetpack Compose, Hilt and other modern practices), working on the NewPipe rewrite might be a good *opportunity* for you, and the team members will be able to provide you with good guidance.

Reviewer 2: ```suggestion
We want to end this paragraph with a **call-to-action**: if you have some time and want to help, feel free to get in touch with us [on the Matrix channel](#conclusion-and-contact) and ask us for something to work on, and **we will give you indications**. When opening a Pull Request, make sure your proposal is not just a Proof-of-Concept, but **provide the best code you can write and highlight the areas that need more work**, so the team can review and merge the PR fast. If you want to learn modern Android development (Kotlin, Jetpack Compose, Hilt and other modern practices), working on the NewPipe rewrite might be a good *opportunity* for you, and the team members will be able to provide you with good guidance.
```

YouTube is requiring integrity checks in their iOS client for most formats returned. We are not able to pass them, so we removed usage of this client in NewPipe. This unfortunately means that livestreams cannot be rewinded for more than 30 seconds, since the *web* client does not allow to rewind livestreams in HLS manifests for more than this duration (while the iOS client allows to rewind up to 1 hour).

The DASH manifests returned by the *web* client are known to be kind of broken and cannot be played directly in the app, but we found a workaround which we should be able to ship in the next release, allowing to rewind up to 4 hours!

All of these durations apply unless the creator disabled rewinding.

## Donations are awesome 

We sincerely **thank all people who have donated** to the NewPipe team and now the [NewPipe e.V.](https://newpipe-ev.de) or helped us in any way, we really appreciate your help and encourage you to continue doing so if you can. You can donate by following the instructions on [our website](https://newpipe.net/donate).

## Conclusion and contact

> In the name of the team, I would like to thank every contributor, triager, translator, tester, donor, and even every user that wrote us kind emails. The community around NewPipe is truly fantastic! If you wish to start contributing, you can reach out to us via Matrix [#newpipe:matrix.newpipe-ev.de](https://matrix.to/#/#newpipe:matrix.newpipe-ev.de) or IRC ([#newpipe on Libera.Chat](https://web.libera.chat/#newpipe)).

Reviewer 2: ```suggestion
In the name of the team, we would like to thank every contributor, triager, translator, tester, donor, and even every user that wrote us kind emails. The community around NewPipe is truly fantastic! If you wish to start contributing, you can reach out to us via Matrix [#newpipe:matrix.newpipe-ev.de](https://matrix.to/#/#newpipe:matrix.newpipe-ev.de) or IRC ([#newpipe on Libera.Chat](https://web.libera.chat/#newpipe)).
```
